File: pantry-play/lib/agent/planner.ts
import { Task, TaskType, TaskPriority } from './types';
import { v4 as uuidv4 } from 'uuid';

export class Planner {
  private tasks: Task[] = [];

  async plan(userInput: string, context: Record<string, any>): Promise<Task[]> {
    const tasks: Task[] = [];
    const input = userInput.toLowerCase();

    // Analyze user intent and create tasks
    if (input.includes('recipe') || input.includes('cook') || input.includes('make')) {
      // First analyze pantry
      tasks.push({
        id: uuidv4(),
        taskType: TaskType.ANALYZE_PANTRY,
        description: 'Analyze available ingredients in pantry',
        priority: TaskPriority.HIGH,
        context: { userInput },
        completed: false,
      });

      // Generate recipe
      tasks.push({
        id: uuidv4(),
        taskType: TaskType.GENERATE_RECIPE,
        description: 'Generate creative recipe based on available ingredients',
        priority: TaskPriority.HIGH,
        context: { 
          preferences: context.dietary_preferences || [],
          servings: context.servings || 2,
        },
        completed: false,
      });

      // Check nutrition if tracking
      if (context.trackNutrition) {
        tasks.push({
          id: uuidv4(),
          taskType: TaskType.CHECK_NUTRITION,
          description: 'Calculate nutritional information',
          priority: TaskPriority.MEDIUM,
          context: {},
          completed: false,
        });
      }
    }

    if (input.includes('shopping') || input.includes('grocery') || input.includes('buy')) {
      tasks.push({
        id: uuidv4(),
        taskType: TaskType.CREATE_SHOPPING_LIST,
        description: 'Generate optimized shopping list',
        priority: TaskPriority.HIGH,
        context: { mealPlan: context.mealPlan || [] },
        completed: false,
      });
    }

    if (input.includes('mood') || input.includes('weather') || input.includes('feeling')) {
      tasks.push({
        id: uuidv4(),
        taskType: TaskType.MOOD_WEATHER_RECOMMEND,
        description: 'Suggest meals based on mood and weather',
        priority: TaskPriority.HIGH,
        context: { 
          mood: context.mood,
          weather: context.weather,
        },
        completed: false,
      });
    }

    if (input.includes('waste') || input.includes('leftover') || input.includes('save')) {
      tasks.push({
        id: uuidv4(),
        taskType: TaskType.TRACK_WASTE,
        description: 'Analyze food waste and suggest leftover transformations',
        priority: TaskPriority.HIGH,
        context: { leftovers: context.leftovers || [] },
        completed: false,
      });
    }

    // Sort by priority
    tasks.sort((a, b) => a.priority - b.priority);
    this.tasks = tasks;
    return tasks;
  }

  getNextTask(): Task | null {
    return this.tasks.find(task => !task.completed) || null;
  }

  markTaskComplete(taskId: string, result: any): void {
    const task = this.tasks.find(t => t.id === taskId);
    if (task) {
      task.completed = true;
      task.result = result;
    }
  }

  getAllTasks(): Task[] {
    return this.tasks;
  }

  getCompletedTasks(): Task[] {
    return this.tasks.filter(task => task.completed);
  }

  getPendingTasks(): Task[] {
    return this.tasks.filter(task => !task.completed);
  }
}

File: pantry-play/src/planner.py
"""
Planner module for Pantry Play
Breaks down user goals into sub-tasks using ReAct pattern
"""

import json
from typing import List, Dict, Any
from enum import Enum

class TaskType(Enum):
    ANALYZE_PANTRY = "analyze_pantry"
    GENERATE_RECIPE = "generate_recipe"
    CHECK_NUTRITION = "check_nutrition"
    SUGGEST_SUBSTITUTIONS = "suggest_substitutions"
    CREATE_SHOPPING_LIST = "create_shopping_list"
    TRACK_WASTE = "track_waste"
    MOOD_WEATHER_RECOMMEND = "mood_weather_recommend"

class TaskPriority(Enum):
    HIGH = 1
    MEDIUM = 2
    LOW = 3

class Task:
    def __init__(self, task_type: TaskType, description: str, priority: TaskPriority, context: Dict[str, Any]):
        self.task_type = task_type
        self.description = description
        self.priority = priority
        self.context = context
        self.completed = False
        self.result = None

class Planner:
    def __init__(self):
        self.tasks = []
        
    def plan(self, user_input: str, context: Dict[str, Any]) -> List[Task]:
        """
        Break down user goals into sub-tasks using ReAct pattern
        """
        tasks = []
        
        # Analyze user intent
        if "recipe" in user_input.lower() or "cook" in user_input.lower():
            # First analyze what's available
            tasks.append(Task(
                TaskType.ANALYZE_PANTRY,
                "Analyze available ingredients in pantry",
                TaskPriority.HIGH,
                {"user_input": user_input}
            ))
            
            # Then generate recipe
            tasks.append(Task(
                TaskType.GENERATE_RECIPE,
                "Generate creative recipe based on available ingredients",
                TaskPriority.HIGH,
                {"preferences": context.get("dietary_preferences", [])}
            ))
            
            # Check nutrition if needed
            if context.get("track_nutrition", False):
                tasks.append(Task(
                    TaskType.CHECK_NUTRITION,
                    "Calculate nutritional information",
                    TaskPriority.MEDIUM,
                    {}
                ))
        
        elif "shopping" in user_input.lower() or "grocery" in user_input.lower():
            tasks.append(Task(
                TaskType.CREATE_SHOPPING_LIST,
                "Generate optimized shopping list",
                TaskPriority.HIGH,
                {"meal_plan": context.get("meal_plan", [])}
            ))
            
        elif "mood" in user_input.lower() or "weather" in user_input.lower():
            tasks.append(Task(
                TaskType.MOOD_WEATHER_RECOMMEND,
                "Suggest meals based on mood and weather",
                TaskPriority.HIGH,
                {"mood": context.get("mood"), "weather": context.get("weather")}
            ))
            
        elif "waste" in user_input.lower() or "leftover" in user_input.lower():
            tasks.append(Task(
                TaskType.TRACK_WASTE,
                "Analyze food waste and suggest leftover transformations",
                TaskPriority.HIGH,
                {"leftovers": context.get("leftovers", [])}
            ))
        
        # Sort by priority
        tasks.sort(key=lambda x: x.priority.value)
        self.tasks = tasks
        return tasks
    
    def get_next_task(self) -> Task:
        """Get the next uncompleted task"""
        for task in self.tasks:
            if not task.completed:
                return task
        return None
    
    def mark_task_complete(self, task: Task, result: Any):
        """Mark a task as complete with its result"""
        task.completed = True
        task.result = result